services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java

```diff
@@ -23,14 +24,27 @@
  *
  * TODO: If later on DDS router could support a better way to register a handler to handle the exceptional cases,
  * we should update the logic here.
- *
- * TODO: Remove all {@link Optional} from this class.
  */
-
 public class RouterExceptionAndTrackingUtils {
   public enum FailureType {
-    REGULAR, SMART_RETRY_ABORTED_BY_SLOW_ROUTE, SMART_RETRY_ABORTED_BY_DELAY_CONSTRAINT,
-    SMART_RETRY_ABORTED_BY_MAX_RETRY_ROUTE_LIMIT, RESOURCE_NOT_FOUND, RETRY_ABORTED_BY_NO_AVAILABLE_REPLICA
+    REGULAR, RESOURCE_NOT_FOUND,
+    SMART_RETRY_ABORTED_BY_SLOW_ROUTE(AggRouterHttpRequestStats::recordSlowRouteAbortedRetryRequest),
+    SMART_RETRY_ABORTED_BY_DELAY_CONSTRAINT(AggRouterHttpRequestStats::recordDelayConstraintAbortedRetryRequest),
+    SMART_RETRY_ABORTED_BY_MAX_RETRY_ROUTE_LIMIT(AggRouterHttpRequestStats::recordRetryRouteLimitAbortedRetryRequest),
+    RETRY_ABORTED_BY_NO_AVAILABLE_REPLICA(AggRouterHttpRequestStats::recordNoAvailableReplicaAbortedRetryRequest);
+
+    final boolean reportUnhealthy;
+    final BiConsumer<AggRouterHttpRequestStats, String> statsRecorder;
+
+    FailureType() {
+      this.reportUnhealthy = true;
+      this.statsRecorder = (stats, store) -> {}; // No op
+    }
+
+    FailureType(BiConsumer<AggRouterHttpRequestStats, String> statsRecorder) {
+      this.reportUnhealthy = false;
+      this.statsRecorder = Objects.requireNonNull(statsRecorder);
+    }
```

Comment:
Very interesting. This makes the below `switch` statement so unnecessary.

Reply:
Yep, adding behavior to enums can be very useful...!